Fix BMW Connected Drive

rikroe opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
The `find_vehicle` service will **always** send the location of your HA instance to BMW. See the [docs](https://www.home-assistant.io/integrations/bmw_connected_drive/#vehicle-finder) for details.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR contains a bugfixes to #59881.

* Upgrade to `bimmer_connected==0.8.5`
  * Release notes:
    * https://github.com/bimmerconnected/bimmer_connected/releases/tag/0.8.3
    * https://github.com/bimmerconnected/bimmer_connected/releases/tag/0.8.4
    * https://github.com/bimmerconnected/bimmer_connected/releases/tag/0.8.5
  * Full diff (most of the commits were due to changing CI to Github actions): https://github.com/bimmerconnected/bimmer_connected/compare/0.8.2...0.8.5
* `find_vehicle` now returns & stores the position of the vehicle, however needs to send a location (using HA location)
* Add check if vehicle tracking is enabled again
* Issues with the BMW API should now be handled more gracefully

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes 
- This PR is related to issue: tracked in library repository
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/20570

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

rikroe committed:
Bump bimmer_connected to 0.8.5

rikroe committed:
Always update HA states after service execution

rikroe committed:
Fix BMW device tracker & vehicle_finder service

rikroe committed:
Add charging_end_time sensor

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/sensor.py:
```diff
@@ -52,6 +54,12 @@ class BMWSensorEntityDescription(SensorEntityDescription):
         unit_metric=TIME_HOURS,
         unit_imperial=TIME_HOURS,
     ),
+    "charging_end_time": BMWSensorEntityDescription(
```
Please separate fixes from feature additions. Add this sensor in a separate PR.

rikroe replied:
Created https://github.com/home-assistant/core/pull/60942 & removed from this PR.

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/sensor.py:
```diff
@@ -171,3 +179,13 @@ def native_value(self) -> StateType:
         return cast(StateType, self.entity_description.value(state, self.hass))
+
+    @property
+    def extra_state_attributes(self) -> Mapping[str, Any] | None:
```
We don't normally want any extra state attributes in sensors. They should be separate sensor entities. Please list what values these attributes are.

rikroe replied:
Moved to #60942 and added an additional sensor

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/__init__.py:
```diff
@@ -117,6 +117,9 @@ def _async_migrate_options_from_data_if_missing(
         options[CONF_READ_ONLY] = data.pop(CONF_READ_ONLY, False)
 
+        # Remove CONF_USE_LOCATION as we have to use it by default
+        options.pop(CONF_USE_LOCATION, None)
```
We shouldn't modify the stored config entry data since we want to be backwards compatible.

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/config_flow.py:
```diff
@@ -115,10 +115,6 @@ async def async_step_account_options(
                         default=self.config_entry.options.get(CONF_READ_ONLY, False),
                     ): bool,
-                    vol.Optional(
-                        CONF_USE_LOCATION,
```
Please remove the corresponding string from strings.json too.

MartinHjelmare replied:
Removing this option could be considered a breaking change.

rikroe replied:
I don't think this is a breaking change, as we have to set `CONF_USE_LOCATION` to `True`.
That is a limitation of the new BMW API, which also requires you to allow using your phone's location to be able to see your vehicle's location. 

However, when thinking about it... We could consider it a breaking change for privacy reasons. In this case, I would opt for keeping the config flow option in and throwing errors/not executing the `find_vehicle` service at all if the option is not set.

MartinHjelmare replied:
I don't think we need the option, but we should make it clear that this has changed by adding a breaking change paragraph in the PR description.

If the user hasn't allowed the location access in the phone, we can raise an error when calling the find vehicle service.

rikroe replied:
Activating location access in the BMW app will not enable it in HA.
As far as I understand the API, you either send a location with the HTTP request for `find_vehicle` and get the location of your vehicle as response, or the vehicle might update some thing in the BMW API but you do not get a direct vehicle location (it might be updated into general status on some cars though).

It is also fine (for more recent cars) if you send a dummy location like the HA default location or (0.0, 0.0) as long as the attributes are given to the API.

MartinHjelmare replied:
What's the use case for reporting your location to BMW servers? It seems it's only interesting to know the location of the vehicle.

rikroe replied:
The problem is BMW wanting to know where you are (whyever they do that). Without sending your `latitude` and `longitude` in the request headers to BMW's servers, the API will respond with `HTTP 400` and the following error message (both have to exist): 
```json
{
    "statusCode": 400,
    "error": "Bad Request",
    "message": "Latitude is required"
}
```

However this is only required if one is actively asking BMW's server, where the vehicle is located. Newer vehicle (at least after 7/2014) also report their location to BMW's server automatically when shutting down the engine. 

If this happens, the `vehicle_finder` service does not need to be used at all as for these vehicles the location data is available in the default status returned by the API.
Older vehicles have to call this service as they will otherwise not get any location data at all.

MartinHjelmare replied:
So is the user locations m only needed when the user uses the `find_vehicle` service? If so, it's enough that we announce this in the breaking change paragraph. Then the users can choose whether they want to use the service or not.

rikroe replied:
Exactly. I have updated the breaking changes paragraph of this PR and also stated it in the docs PR (linked in breaking changes).

MartinHjelmare replied:
Good!

MartinHjelmare replied:
Please just remove the stale string from strings.json and we're good here.

rikroe committed:
Fix pylint & pytest

rikroe committed:
Remove unneeded DEFAULT_OPTION

rikroe committed:
Revert adding charging_end_time & state_attributes

rikroe committed:
Don't delete option data for CONF_USE_LOCATION

rikroe commented:
Removed the deletion of `CONF_USE_LOCATION` from `options` and added a breaking change description that will link to the documentation (of course requiring the docs PR that I just created to be merged as well :) ).

MartinHjelmare commented:
Has the BMW api already changed to require a location? Ie, there's no reason to wait with releasing this breaking change?

rikroe committed:
Remove stale string

rikroe commented:
Just removed the stale string.

And yes, the BMW API has already changed. I just didn't catch everything at once with https://github.com/home-assistant/core/pull/59881.

MartinHjelmare reviewed (approved):
Thanks!

rikroe commented:
Thanks for your support as well!
